### PR TITLE
Making SEB-configuration more granular

### PR DESCRIPTION
**Pull request by Ivanca**
There is 2 new configuration options:
SEB_WHITELIST_PATHS = []
SEB_BLACKLIST_CHAPTERS = []

**SEB_WHITELIST_PATHS** is a list of paths that can be accessed, it may contain one of the following values: 'wiki', 'course-outline', 'courseware', 'progress', 'discussion'; plus any other url name (e.g. useful for custom tabs)

**SEB_BLACKLIST_CHAPTERS** is a list of chapters that are blocked even when 'courseware' has been white-listed; e.g. useful to block the "exam" chapter but not the rest.

Right now those 2 configurations are global and not in a course-basis; may be better to agree on the current state before making it so.

### CODE REVIEW

**Comment by Ivanca** [comment]
@felipemontoya feedback pending

File: seb_openedx/edxapp_wrapper/backends/edxmako_module_h_v1.py [hunk 1]
```diff
@@ -1,10 +1,15 @@
 """ Backend abstraction """
 import os
 import edxmako  # pylint: disable=import-error
+if os.path.dirname(os.path.abspath(__file__)) + '/templates' not in edxmako.LOOKUP['main'].directories:
```
Comment (felipemontoya) [inline seb_openedx/edxapp_wrapper/backends/edxmako_module_h_v1.py:4]: This here will make sure that we only include this once, correct?

File: seb_openedx/middleware.py [hunk 2]
```diff
@@ -1,27 +1,132 @@
...
+            if is_courseware_view:
+                return self.courseware_error_response(request, *view_args, **view_kwargs)
+            return self.generic_error_response(request, course_key)
+        return None
+
+    def is_whitelisted_view(self, request, course_key):
+        """ First broad filter: whitelisting of paths/tabs """
+        views_module = inspect.getmodule(request.resolver_match.func).__name__
+
+        if 'discussion' in SEB_WHITELIST_PATHS and views_module == 'discussion.views':
```
Comment (felipemontoya) [inline seb_openedx/middleware.py]: What do you think about making this check if `views_module in SEB_WHITELIST_PATHS` and per default use `discussion.views` in the list?

File: seb_openedx/middleware.py [hunk 3]
```diff
@@ -1,27 +1,132 @@
...
+        chapter = request.resolver_match.kwargs.get('chapter')
+
+        if not SEB_BLACKLIST_CHAPTERS:
+            return False
+
+        if chapter in SEB_BLACKLIST_CHAPTERS:
+            return True
+
+        if 'courseware' in SEB_WHITELIST_PATHS and self.is_xblock_request(request):
+            modulestore = get_xmodule_modulestore_module().django.modulestore
```
Comment (felipemontoya) [inline seb_openedx/middleware.py]: If all the reason to create the `get_xmodule_modulestore_module` function is to get the `path` on line 99, why not create a function `get_path_to_location` in the edxapp_wrapper and abstract all that code from here?

Comment (felipemontoya) [reply seb_openedx/middleware.py]: Even better, `get_chapter_from_location` since that is the real reason for all this

File: seb_openedx/middleware.py [hunk 4]
```diff
@@ -1,27 +1,132 @@
...
+            path_to_location = get_xmodule_modulestore_module().search.path_to_location
+            usage_id = request.resolver_match.kwargs.get('usage_id')
+            usage_key = UsageKey.from_string(unquote_slashes(usage_id)).map_into_course(course_key)
+            if usage_key:
+                path = path_to_location(modulestore(), usage_key)
+                chapter = path[1]
```
Comment (felipemontoya) [inline seb_openedx/middleware.py]: The 1 is acting as a Magic Number. Can we refactor that?

File: seb_openedx/middleware.py (hunk 2 above)
Comment (felipemontoya) [reply seb_openedx/middleware.py]: This complete function has a very high complexity derived from the likely to grow ammount of if statements. I think we can refactor it a bit so that most of what it does is straight forward compare some variables to its existence on the WHITELIST_PATHS list

File: seb_openedx/middleware.py [hunk 5]
```diff
@@ -1,27 +1,132 @@
...
+        context = {
+            'course': course,
+            'request': request
+        }
+
+        return render_to_response('seb-403.html', context, status=403)
+
+    def is_xblock_request(self, request):
+        """ returns if it's an xblock HTTP request or not """
+        return request.resolver_match.func.__name__ == 'handle_xblock_callback'
```
Comment (felipemontoya) [inline seb_openedx/middleware.py:132]: This returns true only for requests of the form: `/courses/<ID>/xblock/block-v1:<ID>+type@type_of_xblock+block@<SOMEHASH>/handler/name_of_handler` or are we catching something else?

File: seb_openedx/seb_courseware_index.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    """ Class to extend Open edX class CoursewareIndex and change _create_courseware_context """
+    fragment = None
+
+    @classmethod
+    def set_context_fragment(cls, new_fragment):
+        """ set for later usage at context["fragment"] """
+        cls.fragment = new_fragment
+
+    def _create_courseware_context(self, *args, **kwargs):
+        """ ovrriding method (but still calling the overwritten one) """
```
Comment (felipemontoya) [inline seb_openedx/seb_courseware_index.py]:
```suggestion
        """ Overriding method (but still calling the overwritten one) """
```

File: seb_openedx/settings/common.py [hunk 7]
```diff
@@ -29,7 +29,17 @@ def plugin_settings(settings):
...
     settings.SEB_CONFIGURATION_HELPERS = 'seb_openedx.edxapp_wrapper.backends.get_configuration_helpers_h_v1'
     settings.SEB_EDXMAKO_MODULE = 'seb_openedx.edxapp_wrapper.backends.edxmako_module_h_v1'
     settings.SEB_PERMISSION_COMPONENTS = ['AlwaysAllowStaff', 'CheckSEBKeysRequestHash']
     settings.SEB_KEY_SOURCES = ['from_global_settings', 'from_other_course_settings', 'from_site_configuration']
+
+    # When SEB determines the access is denied one may specify what to whitelist/blacklist more granularly
+    settings.SEB_WHITELIST_PATHS = []
+    settings.SEB_BLACKLIST_CHAPTERS = []
+
+    # Example allowing all tabs but denying access to a specific "chapter":
```
Comment (felipemontoya) [inline seb_openedx/settings/common.py]: Lets better put thin in the readme, not in the actual settings

File: seb_openedx/settings/common.py [hunk 8]
```diff
@@ -29,7 +29,17 @@ def plugin_settings(settings):
...
+    settings.SEB_COURSEWARE_MODULE = 'seb_openedx.edxapp_wrapper.backends.get_courseware_module_h_v1'
+    settings.SEB_XMODULE_MODULESTORE_MODULE = 'seb_openedx.edxapp_wrapper.backends.get_courseware_module_h_v1'
     settings.SEB_CONFIGURATION_HELPERS = 'seb_openedx.edxapp_wrapper.backends.get_configuration_helpers_h_v1'
     settings.SEB_EDXMAKO_MODULE = 'seb_openedx.edxapp_wrapper.backends.edxmako_module_h_v1'
     settings.SEB_PERMISSION_COMPONENTS = ['AlwaysAllowStaff', 'CheckSEBKeysRequestHash']
     settings.SEB_KEY_SOURCES = ['from_global_settings', 'from_other_course_settings', 'from_site_configuration']
+
+    # When SEB determines the access is denied one may specify what to whitelist/blacklist more granularly
+    settings.SEB_WHITELIST_PATHS = []
+    settings.SEB_BLACKLIST_CHAPTERS = []
```
Comment (felipemontoya) [inline seb_openedx/settings/common.py:41]: At some point this will be defined in the UI for each course, both SEB_WHITELIST_PATHS and SEB_BLACKLIST_CHAPTERS, correct?

File: seb_openedx/templates/seb-403-error-message.html [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+from django.utils.translation import ugettext as _
+from openedx.core.djangolib.markup import Text
+
+%>
+<h1><%block name="pageheader">${_("Error: Access not allowed")}</%block></h1>
+<p style="text-align: center;">
+    <%block name="pagecontent">
+        ${Text(_(' This course can only be accessed with "Safe Exam Browser"'))}
+    </%block>
+</p>
```
Comment (felipemontoya) [inline seb_openedx/templates/seb-403-error-message.html]:
```suggestion
</p>

```

Comment (felipemontoya) [reply seb_openedx/templates/seb-403-error-message.html]: empty line at the end

File: seb_openedx/tests/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+def get_courseware_module():
+    """ helper for settings.test fake backend """
+    class Object(object):
+        """ dummy class """
+        pass
+
+    def get_course(course_id, depth=0, **kwargs):
+        """ dummy get_course function """
+        return None
+    fake_courseware_module = Object()
```
Comment (felipemontoya) [inline seb_openedx/tests/test_utils.py]: Would a MagicMock not do all this by default?

Comment (felipemontoya) [reply seb_openedx/tests/test_utils.py]: then you only need to tell it to patch `fake_courseware_module.courses.get_course = get_course`

File: seb_openedx/utils.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
+""" utils """
```
Comment (felipemontoya) [inline seb_openedx/utils.py]: If we do the refactor for get_chapter_from_location, this whole file can go and use the original edx-platform/openedx/core/lib/url_utils.py functions

**Review by felipemontoya — CHANGES_REQUESTED** [review]

File: seb_openedx/edxapp_wrapper/backends/edxmako_module_h_v1.py (hunk 1 above)
Comment (Ivanca) [reply seb_openedx/edxapp_wrapper/backends/edxmako_module_h_v1.py:4]: yep

File: seb_openedx/middleware.py (hunk 2 above)
Comment (Ivanca) [reply seb_openedx/middleware.py]: > What do you think about making this check if `views_module in SEB_WHITELIST_PATHS` and per default use `discussion.views` in the list?

Sure

> This complete function has a very high complexity derived from the likely to grow ammount of if statements. I think we can refactor it a bit so that most of what it does is straight forward compare some variables to its existence on the WHITELIST_PATHS list

The mental complexity is low due early returns, meaning any of those paths from line 56 to 68 end the function and the rest of the functions doesn't have to be understood if it stops there: and when it doesn't stop there then those lines don't matter and the logic happens between line 69 through 81

File: seb_openedx/middleware.py (hunk 3 above)
Comment (Ivanca) [reply seb_openedx/middleware.py]: Ok

File: seb_openedx/middleware.py (hunk 4 above)
Comment (Ivanca) [reply seb_openedx/middleware.py]: I guess we can assign 1 to a variable called chapter_index and then use that variable (beyond that not much is possible without using reflection to check what argument returns the chapter)

File: seb_openedx/middleware.py (hunk 5 above)
Comment (Ivanca) [reply seb_openedx/middleware.py:132]: I think that's all we need to match for xblocks

File: seb_openedx/settings/common.py (hunk 8 above)
Comment (Ivanca) [reply seb_openedx/settings/common.py:41]: Yep

File: seb_openedx/settings/common.py (hunk 7 above)
Comment (Ivanca) [reply seb_openedx/settings/common.py]: Sure, but I think Edx-plat would be much less painful if each setting had a comment over it about what it does and a few examples.

File: seb_openedx/tests/test_utils.py (hunk 10 above)
Comment (Ivanca) [reply seb_openedx/tests/test_utils.py]: Ok

File: seb_openedx/utils.py (hunk 11 above)
Comment (Ivanca) [reply seb_openedx/utils.py]: Ok

File: seb_openedx/middleware.py (hunk 2 above)
Comment (felipemontoya) [reply seb_openedx/middleware.py]: For the second comment, the current complexity is not skyhigh, but I do see this becoming a problem when there are new `if`s and more checks. Which is very likely to happen after new edge cases are revealed during a long test run

File: seb_openedx/middleware.py (hunk 4 above)
Comment (felipemontoya) [reply seb_openedx/middleware.py]: That is the only way to fix magic numbers

File: seb_openedx/middleware.py (hunk 2 above)
Comment (Ivanca) [reply seb_openedx/middleware.py]: A lot of the logic here cannot be converted into simple strings in whitelist_path, for example `if 'courseware' in SEB_WHITELIST_PATHS and self.is_xblock_request(request):`is only for xblock requests made on the courseware.

The other thing is that making the user type hard to remember words like 'course_wiki.views' (instead of just `wiki`) is anti-usability; and since the abstraction is a level higher if tomorrow e.g. edx changes the name for 
"course_wiki_media" the abstraction would map "wiki" to that new key, instead of asking every user to change every place they wrote "course_wiki.views"

Comment (felipemontoya) [reply seb_openedx/middleware.py]: If you are worried about the mapping, we can make a mapper dictionary and still make this method a lot simpler. Also, the part of the courseware I understand why it needs to be different, but we can make all that with 3 checks tops and also include the possibility to grow in the future. Right now we have 5 checks and no easy extension capability

Comment (Ivanca) [reply seb_openedx/middleware.py]: the easy extension is using url_name, for example if you want to block the following urls:
 ```python
[
  url(r'^course_specific_login/{}/$'.format(settings.COURSE_ID_PATTERN),
  external_auth_views.course_specific_login, name='course-specific-login'),
  url(r'^course_specific_register/{}/$'.format(settings.COURSE_ID_PATTERN),
  external_auth_views.course_specific_register, name='course-specific-register'),
]` 
```
all you have to do is add "course-specific-" to SEB_WHITELIST_PATHS

The reason the other "ifs" are needed is because a lot of the urls we need to blocks don't have a name (or needs complex logic)

Comment (felipemontoya) [reply seb_openedx/middleware.py]: I think that is very good, that is why I think only 3 checks are enough.
1 for courseware
2 for the urls
3 for the mapper dict
That would be brilliant

**Comment by Ivanca** [comment]
@felipemontoya https://github.com/eduNEXT/seb-openedx/pull/12/commits/57e4fc697a5e911b264b5ffd19bf50b0ff36b713 Commit addressing feedback

File: seb_openedx/edxapp_wrapper/backends/get_chapter_from_location_h_v1.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+from xmodule.modulestore.django import modulestore  # pylint: disable=import-error
+from xmodule.search import path_to_location  # pylint: disable=import-error
+
+
+def get_chapter_from_location(usage_id, course_key):
+    """ hawthorn backend """
+    usage_key = UsageKey.from_string(unquote_slashes(usage_id)).map_into_course(course_key)
+    if usage_key:
+        path = path_to_location(modulestore(), usage_key)
+        chapter_index = 1
```
Comment (felipemontoya) [inline seb_openedx/edxapp_wrapper/backends/get_chapter_from_location_h_v1.py]: Would you like making this a CHAPTER_INDEX "constant" ?

File: seb_openedx/middleware.py [hunk 13]
```diff
@@ -1,27 +1,130 @@
...
+        html = Fragment()
+        html.add_content(render_to_string('seb-403-error-message.html', {}))
+        SebCoursewareIndex.set_context_fragment(html)
+        return SebCoursewareIndex.as_view()(request, *view_args, **view_kwargs)
+
+    def generic_error_response(self, request, course_key):
+        """ generic error response, full page 403 error (with course menu) """
+        courseware = get_courseware_module()
+        try:
+            course = courseware.courses.get_course(course_key, depth=2)
```
Comment (felipemontoya) [inline seb_openedx/middleware.py:119]: Would it make sense to wrapp `get_course` instead of `get_courseware_module` ?

File: seb_openedx/middleware.py [hunk 14]
```diff
@@ -1,27 +1,130 @@
...
+                access_denied = True
+
             active_comps = get_enabled_permission_classes()
             for permission in active_comps:
                 if permission().check(request, course_key):
-                    return
-            return render_to_response('seb-403.html', status=403)
+                    access_denied = False
+
+        if access_denied:
```
Comment (felipemontoya) [inline seb_openedx/middleware.py:46]: this part, where we answer once we have decided that the view needs to be denied, could be refactored into a function. Here we will be doing more things like putting a permanent mark on the user or reporting the issue somewhere

**Review by felipemontoya — CHANGES_REQUESTED** [review]
Minor things. It looks great

File: seb_openedx/middleware.py (hunk 13 above)
Comment (Ivanca) [reply seb_openedx/middleware.py:119]: We use get_courseware_module on another file as well: seb_courseware_index.py
And we probably gonna be using some attributes (and deeper ones) of the courseware_module in the near future, so having separate backends for each function will result in too many files.

File: seb_openedx/middleware.py (hunk 14 above)
Comment (Ivanca) [reply seb_openedx/middleware.py:46]: Ok

File: seb_openedx/edxapp_wrapper/backends/get_chapter_from_location_h_v1.py (hunk 12 above)
Comment (Ivanca) [reply seb_openedx/edxapp_wrapper/backends/get_chapter_from_location_h_v1.py]: Ok

**Review by felipemontoya — APPROVED** [review]
@Ivanca just address the feedback given per other channels
